### DISCUSSION TITLE
Need a way to encrypt OTP seeds with say java keystore private key

### DISCUSSION THREAD

**Discussion by navatweb** [keycloak/keycloak#35746] [Ideas]
Currently One Time Password seeds are stored essentially in clear in the SQL database (base32 but basically in clear).
For enterprise systems, important secrets such as OTP seeds must be stored encrypted.
There is a way to use the private key from the external java keystore to issues tokens. It could be an option to use these keys to encrypt the seeds. But as such keys expire, the seeds encrypted with an expiring key must be reencrypted. Among other things I have not thought through.

**Comment by stianst** [comment]
Database encryption is an option here, but I appreciate that is a bit of a hammer approach and not trivial.

Problem with introducing some sort of application side encryption of specific values aka secrets is key management and especially rotation of keys, or switching encryption algorithms. Especially in the case where an on-demand scenario isn't sufficient, which would be the case for OTP secrets as you couldn't keep the old keys around until all users have eventually logged-in, but would need to somehow eagerly re-encrypt all secrets.

**Comment by navatweb** [comment]
Thanks [stianst](https://github.com/stianst) for the quick review.
We already encrypt the whole database but seeds are still vulnerable to SQL injection and other attacks.

Yes, anytime active key changes, all OTP seeds in the realm will have to be re-encrypted. Probably not a big deal unless you have a huge number of users in the realm.

Using the private key stored inside the java keystore will be a good candidate for use in encryption, as KeyCloak already supports storing keys outside of the database in this fashion.
